### PR TITLE
feature/ECUBattery/ SyncGenerateFrames and Tests

### PR DESCRIPTION
**Pull request by RubenNatanael**
Sync with GenerateFrame from MCU:

- Adding option to choose type of frame(data or remote)
- Generate coverage files using lcov and gcov
- Adding new test to improve code cover

### CODE REVIEW

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 1]
```diff
@@ -36,484 +36,475 @@
...
+        8:  187:        for (int i = 0; i < 3; i++)
+        -:  188:        {
+        6:  189:            data.push_back(response[i]);
+        -:  190:        }
+        2:  191:        this->sendFrame(id, data);
+        2:  192:        return;
+        -:  193:    }
+        -:  194:    else
+        -:  195:    {
+        -:  196:        /*Delete first 3 data that were sended in the first frame */
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:200]: Add space after /*

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 2]
```diff
@@ -36,484 +36,475 @@
...
+        2:  216:}
+        -:  217:
+        1:  218:void GenerateFrames::flowControlFrame(int id)
+        -:  219:{
+        1:  220:    this->sendFrame(id, {0x30,0x00,0x00,0x00});
+        1:  221:}
+        -:  222:
+        2:  223:void GenerateFrames::readMemoryByAddress(int id, int memory_size, int memory_address, std::vector<int> response )
+        -:  224:{
+        -:  225:    /*add lengths of of memory size/address to the frame*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:229]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 3]
```diff
@@ -36,484 +36,475 @@
...
+        2:  226:    int length_memory_size = (countDigits(memory_size) +1) / 2;
+        2:  227:    int length_memory_address = (countDigits(memory_address) + 1) / 2;
+        2:  228:    int length_memory = length_memory_size * 0x10 + length_memory_address;
+        2:  229:    if (response.size() == 0)
+        -:  230:    {
+        1:  231:        int pci_l = length_memory_size + length_memory_address + 2;
+        1:  232:        std::vector<int> data = {pci_l, 0x23, length_memory};
+        -:  233:        /*add memory address and size to the frame*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:237]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 4]
```diff
@@ -36,484 +36,475 @@
...
-        -:  514:    }
-        9:  515:}
+    #####:  254:        std::cout<<"ERROR: Response to long, consider using ReadMemoryByAdressLongResponse method\n";
+    #####:  255:        return;
+        -:  256:    }
+        -:  257:}
+        -:  258:
+        2:  259:void GenerateFrames::readMemoryByAddressLongResponse(int id, int memory_size, int memory_address, std::vector<int> response, bool first_frame)
+        -:  260:{
+        -:  261:    /*add lengths of of memory size/address to the frame*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:265]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 5]
```diff
@@ -36,484 +36,475 @@
...
+        2:  262:    int length_memory_size = (countDigits(memory_size) +1) / 2;
+        2:  263:    int length_memory_address = (countDigits(memory_address) + 1) / 2;
+        2:  264:    int length_memory = length_memory_size * 0x10 + length_memory_address;
+        2:  265:    if (first_frame)
+        -:  266:    {
+        1:  267:        int pci_l = (int)response.size() + 2 + length_memory_size + length_memory_address;
+        1:  268:        std::vector<int> data = {0x10, pci_l, 0x63, length_memory};
+        -:  269:        /*add memory address and size to the frame*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:273]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 6]
```diff
@@ -36,484 +36,475 @@
...
+        1:  270:        insertBytes(data, memory_address, length_memory_address);
+        1:  271:        insertBytes(data, memory_size, length_memory_size);
+        -:  272:        /*Send only 3 first bytes of data*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:276]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 7]
```diff
@@ -36,484 +36,475 @@
...
+        -:  274:        {
+        1:  275:            data.push_back(response[i]);
+        -:  276:        }
+        1:  277:        this->sendFrame(id, data);
+        1:  278:        return;
+        -:  279:    }
+        -:  280:    else
+        -:  281:    {
+        1:  282:        int memory_allready_send = 8 - (length_memory_address + length_memory_size + 4);
+        -:  283:        /*Delete data allready sended*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:287]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 8]
```diff
@@ -36,484 +36,475 @@
...
+        1:  359:            this->sendFrame(id, data);
+        1:  360:            return;
+        -:  361:        } else
+        -:  362:        {
+    #####:  363:            std::cout<<"ERROR: Can't send more than 7 DTC/frame, please consider send 2 or more frames\n";
+    #####:  364:            return;
+        -:  365:        }
+        -:  366:        
+        -:  367:    }
+        -:  368:    /*Response*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:372]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 9]
```diff
@@ -36,484 +36,475 @@
...
+        1:  387:void GenerateFrames::negativeResponse(int id, int sid, int nrc)
+        -:  388:{
+        2:  389:    std::vector<int> data = {0x03, 0x7F, sid, nrc};
+        1:  390:    this->sendFrame(id, data);
+        2:  391:    return;
+        -:  392:}
+        -:  393:
+        1:  394:void GenerateFrames::requestDownload(int id, int data_format_identifier, int memory_address, int memory_size)
+        -:  395:{
+        -:  396:    /* Request Frame
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:400]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 10]
```diff
@@ -36,484 +36,475 @@
...
+        -:  403:    /*add memory address and size to the frame*/
+        1:  404:    insertBytes(data, memory_address, length_memory_address);
+        1:  405:    insertBytes(data, memory_size, length_memory_size);
+        1:  406:    this->sendFrame(id, data);
+        2:  407:    return;
+        -:  408:}
+        -:  409:
+        1:  410:void GenerateFrames::requestDownloadResponse(int id, int max_number_block)
+        -:  411:{
+        -:  412:    /*Response frame*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:416]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 11]
```diff
@@ -36,484 +36,475 @@
...
+        1:  394:void GenerateFrames::requestDownload(int id, int data_format_identifier, int memory_address, int memory_size)
+        -:  395:{
+        -:  396:    /* Request Frame
+        -:  397:    add lengths of of memory size/address to the frame*/
+        1:  398:    int length_memory_size = countDigits(memory_size + 1) / 2;
+        1:  399:    int length_memory_address = countDigits(memory_address + 1) / 2;
+        1:  400:    int length_memory = length_memory_size * 0x10 + length_memory_address;
+        1:  401:    int pci_length = length_memory_size + length_memory_address + 3; //pci_l
+        2:  402:    std::vector<int> data = {pci_length, 0x34, data_format_identifier, length_memory}; //pci_l
+        -:  403:    /*add memory address and size to the frame*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:407]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 12]
```diff
@@ -36,484 +36,475 @@
...
+        1:  413:    int length_max_number_block = (countDigits(max_number_block) + 1) / 2;
+        2:  414:    std::vector<int> data = {length_max_number_block + 2, 0x74, (length_max_number_block * 0x10)}; //pci_l
+        1:  415:    insertBytes(data, max_number_block, length_max_number_block);
+        1:  416:    this->sendFrame(id, data);
+        2:  417:    return;
+        -:  418:}
+        -:  419:
+        3:  420:void GenerateFrames::transferData(int id, int block_sequence_counter, std::vector<int> transfer_request)
+        -:  421:{
+        -:  422:    /*If is not a response*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:426]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 13]
```diff
@@ -36,484 +36,475 @@
...
+        5:  444:        for (int i = 0; i < 4 ; i++)
+        -:  445:        {
+        4:  446:            data.push_back(transfer_request[i]);
+        -:  447:        }
+        1:  448:        this->sendFrame(id, data);
+        1:  449:        return;
+        -:  450:    }
+        -:  451:    else
+        -:  452:    {
+        -:  453:        /*Delete first 3 data that were sended in the first frame*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:457]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 14]
```diff
@@ -36,484 +36,475 @@
...
+        1:  476:        return;
+        -:  477:    }
+        2:  478:    std::vector<int> data = {0x01,0x77};
+        1:  479:    this->sendFrame(id, data);
+        1:  480:    return;
+        -:  481:}
+        -:  482:
+    #####:  483:bool GenerateFrames::requestUpdateStatus(int id, bool response)
+        -:  484:{
+        -:  485:    /*No impplementation, I don't find this service in the standart*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:489]: same here

File: src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov [hunk 15]
```diff
@@ -36,484 +36,475 @@
...
+    #####:  486:    return false;
+        -:  487:}
+        -:  488:
+        -:  489:/*Private*/
```
Comment (Miruna-Maria) [inline src/ecu_simulation/BatteryModule/test/GenerateFrames.cpp.gcov:493]: same here

**Review by Miruna-Maria — APPROVED** [review]

**Review by GeorgeCimpoeru — APPROVED** [review]

**Pull request merged**